dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageProblemCauseAnnotator.java

```diff
@@ -0,0 +1,122 @@
+      ClassFile sourceClass = linkageProblem.getSourceClass();
+      ClassPathEntry sourceEntry = sourceClass.getClassPathEntry();
+
+      Artifact sourceArtifact = sourceEntry.getArtifact();
+
+      // Resolves the dependency graph with the source artifact at the root.
+      ClassPathBuilder classPathBuilder = new ClassPathBuilder();
+      ClassPathResult subtreeResult = classPathBuilder.resolveWithMaven(sourceArtifact);
+
+      ClassPathEntry entryInSubtree =
+          findClassPathEntryForSymbol(subtreeResult, linkageProblem.getSymbol());
+      if (entryInSubtree == null) {
+        linkageProblem.setCause(UnknownCause.getInstance());
+      } else {
+        Artifact artifactInSubtree = entryInSubtree.getArtifact();
+        DependencyPath pathToSourceEntry = rootResult.getDependencyPaths(sourceEntry).get(0);
+        DependencyPath pathFromSourceEntryToUnselectedEntry =
+            subtreeResult.getDependencyPaths(entryInSubtree).get(0);
+        DependencyPath pathToUnselectedEntry =
+            pathToSourceEntry.concat(pathFromSourceEntryToUnselectedEntry);
+
+        ClassPathEntry selectedEntry =
+            findEntryByArtifactId(
+                rootResult, artifactInSubtree.getGroupId(), artifactInSubtree.getArtifactId());
+        if (selectedEntry != null) {
+          Artifact selectedArtifact = selectedEntry.getArtifact();
+          if (!selectedArtifact.getVersion().equals(artifactInSubtree.getVersion())) {
+            // Different version of that artifact is selected in rootResult
+            linkageProblem.setCause(
+                new DependencyConflict(
+                    rootResult.getDependencyPaths(selectedEntry).get(0), pathToUnselectedEntry));
+          } else {
+            // A linkage error was already there when sourceArtifact was built.
+            linkageProblem.setCause(UnknownCause.getInstance());
+          }
+        } else {
+          // No artifact that matches groupId and artifactId in rootResult.
+          linkageProblem.setCause(new MissingDependency(pathToUnselectedEntry));
+        }
+      }
+    }
+  }
+
+  /**
+   * Returns the class path entry in {@code classPathResult} for the artifact that matches {@code
+   * groupId} and {@code artifactId}. {@code Null} if no matching artifact is found.
+   */
+  private static ClassPathEntry findEntryByArtifactId(
+      ClassPathResult classPathResult, String groupId, String artifactId) {
+    for (ClassPathEntry entry : classPathResult.getClassPath()) {
+      Artifact artifact = entry.getArtifact();
+      if (artifact.getGroupId().equals(groupId) && artifact.getArtifactId().equals(artifactId)) {
+        return entry;
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Returns the {@link ClassPathEntry} in {@code classPathResult} that contains the class of {@code
+   * symbol}. {@code Null} if no matching entry is found.
+   */
+  private static ClassPathEntry findClassPathEntryForSymbol(
```

Comment:
This also feels like it fits naturally as an instance method in ClasspathResult

Reply:
Yes, moved.